apps/prairielearn/src/sprocs/check_assessment_access_rule.sql

```diff
@@ -41,6 +42,17 @@ BEGIN
 
     << prairietest_access >>
```

Comment:
I believe we could replace the `<<prairietest_access>>` block with the following, to obtain identical results with hopefully clearer control flow?
```
IF (exam_uuid IS NOT NULL AND mode != 'Exam') THEN
    -- only allowed to use exam_uuid in mode Exam
    authorized := FALSE;
END IF;

IF (exam_uuid IS NOT NULL AND mode_reason != 'PrairieTest') THEN
    -- only allowed to use exam_uuid with PT
    authorized := FALSE;
END IF;

IF (mode_reason = 'PrairieTest') THEN
    -- we must have `mode = 'Exam'` here because this is guaranteed by `mode_reason = 'PrairieTest'`

    -- is there a checked-in pt_reservation for the (possibly-NULL) exam_uuid?
    -- SAME QUERY AS WE ALREADY HAVE FOR THIS
    -- if exam_uuid IS NULL then we definitely won't find a reservation

    IF NOT FOUND THEN
        authorized := FALSE;
    END IF;
END IF;
```

Reply:
Perhaps this is solely based on vibes and the query planner will actually do the right thing, but it feels like we could save some work by skipping the reservation query entirely if there's no `exam_uuid`? Or maybe the query planner will see that we're trying to find a `NULL` value on a column with a `NON NULL` constraint and just immediately return zero rows? 🤔

Regardless, I like the idea of getting rid of `<< prairietest_access >>` and `EXIT prairietest_access`, which we can still do whether or not we decide we can skip the `exam_uuid IS NOT NULL` check.

Reply:
I'm fine to have an explicit check for `exam_uuid IS NOT NULL`, especially considering that it will serve as documentation for what we are doing.

Reply:
I was ultimately able to get the tests to pass with this:

```sql
    IF (assessment_access_rule.exam_uuid IS NULL AND mode = 'Exam') THEN
        -- Public mode is required for this access rule.
        authorized := FALSE;
    END IF;

    IF (assessment_access_rule.exam_uuid IS NOT NULL AND mode IS DISTINCT FROM 'Exam') THEN
        -- Exam mode is required for this access rule.
        authorized := FALSE;
    END IF;

    -- IF (assessment_access_rule.exam_uuid IS NOT NULL AND mode_reason != 'PrairieTest') THEN
    --     -- PrairieTest is required for this access rule.
    --     authorized := FALSE;
    -- END IF;

    IF (assessment_access_rule.exam_uuid IS NOT NULL AND mode = 'Exam' AND mode_reason = 'PrairieTest') THEN
        -- Look for a checked-in PrairieTest reservation.
        SELECT r.access_end
        INTO exam_access_end
        FROM
            pt_reservations AS r
            JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
            JOIN pt_exams AS x ON (x.id = r.exam_id)
        WHERE
            (date BETWEEN r.access_start AND r.access_end)
            AND e.user_id = check_assessment_access_rule.user_id
            AND x.uuid = assessment_access_rule.exam_uuid;

        IF NOT FOUND THEN
            -- No reservation found; block access.
            authorized := FALSE;
        END IF;
    END IF;
```

Note how one of the checks is commented out, and I added another check at the beginning. The commented-out check had no bearing on the tests passing. So, either it's truly unnecessary, or we're missing a test.

_Or_, looking at your other comments, we're passing `authz_mode_reason: 'PrairieTest'` where we shouldn't be.

Reply:
I like your new version of the code. Much cleaner!

I was getting too confused trying to think through all the cases so I ended up writing out all the legal variants of variables in the big table below.

## Proposed code changes

If we care about the cases identified in the `comments` column as possibly wanting a change then I think we could fix them by adding:

```
IF (assessment_access_rule.exam_uuid IS NOT NULL AND mode_reason IS DISTINCT FROM 'PrairieTest') THEN
    -- only use exam_uuid when we are using PrairieTest
    authorized := FALSE;
END IF;

IF (assessment_access_rule.exam_uuid IS NOT NULL AND assessment_access_rule.mode IS DISTINCT FROM 'Exam') THEN
    -- only use exam_uuid if we have an explicit mode=Exam set in the access rule
    authorized := FALSE;
END IF;
```

## Definitions of columns and values

The first columns are the values in the `assessment_access_rule` and arguments, while the later columns are:
* `reservation`: Does the SELECT for a `pt_reservation` find one?
* `authz`: This is what I think the `<< prairietest_access >>` block should be returning.
* `old authz`: This is what the old code block used to return.
* `new authz`: This is what the current new code (as of https://github.com/PrairieLearn/PrairieLearn/pull/9743/commits/41632fc2bf3b842c4e865a68726646723f9f7cc9) returns.

For the `authz` values I listed `FALSE` if the `<< prairietest_access >>` block set `authorized := FALSE`, and `TRUE` otherwise. A value of `TRUE` here really just means that `authorized` is not modified by the `<< prairietest_access >>` block, not that it is explicitly set to `TRUE`.

Note the `comments` column at the end.

## The big table of variables

```
aar.mode | aar.exam_uuid | mode   | mode_reason | reservation | authz | old authz | new authz | comment
---------|---------------|--------|-------------|-------------|-------|-----------|-----------|--------
Exam     | NULL          | Exam   | NULL        | FOUND       | ---   | ---       | ---       |
Exam     | NULL          | Exam   | NULL        | NOT FOUND   | TRUE  | TRUE      | TRUE      |
Exam     | NULL          | Exam   | PrairieTest | FOUND       | ---   | ---       | ---       |
Exam     | NULL          | Exam   | PrairieTest | NOT FOUND   | FALSE | TRUE      | FALSE     | Intended change for this PR
Exam     | NULL          | Exam   | Other       | FOUND       | ---   | ---       | ---       |
Exam     | NULL          | Exam   | Other       | NOT FOUND   | TRUE  | TRUE      | TRUE      |
Exam     | not-NULL      | Exam   | NULL        | FOUND       | FALSE | TRUE      | TRUE      | Do we want a change here?
Exam     | not-NULL      | Exam   | NULL        | NOT FOUND   | FALSE | FALSE     | TRUE      | Unintended change?
Exam     | not-NULL      | Exam   | PrairieTest | FOUND       | TRUE  | TRUE      | TRUE      |
Exam     | not-NULL      | Exam   | PrairieTest | NOT FOUND   | FALSE | FALSE     | FALSE     |
Exam     | not-NULL      | Exam   | Other       | FOUND       | FALSE | TRUE      | TRUE      | Do we want a change here?
Exam     | not-NULL      | Exam   | Other       | NOT FOUND   | FALSE | FALSE     | TRUE      | Unintended change?
Public   | NULL          | Public | NULL        | FOUND       | ---   | ---       | ---       |
Public   | NULL          | Public | NULL        | NOT FOUND   | TRUE  | TRUE      | TRUE      |
Public   | NULL          | Public | PrairieTest | FOUND       | ---   | ---       | ---       |
Public   | NULL          | Public | PrairieTest | NOT FOUND   | ---   | ---       | ---       |
Public   | NULL          | Public | Other       | FOUND       | ---   | ---       | ---       |
Public   | NULL          | Public | Other       | NOT FOUND   | TRUE  | TRUE      | TRUE      |
Public   | not-NULL      | Public | NULL        | FOUND       | FALSE | FALSE     | FALSE     |
Public   | not-NULL      | Public | NULL        | NOT FOUND   | FALSE | FALSE     | FALSE     |
Public   | not-NULL      | Public | PrairieTest | FOUND       | FALSE | FALSE     | FALSE     |
Public   | not-NULL      | Public | PrairieTest | NOT FOUND   | FALSE | FALSE     | FALSE     |
Public   | not-NULL      | Public | Other       | FOUND       | FALSE | FALSE     | FALSE     |
Public   | not-NULL      | Public | Other       | NOT FOUND   | FALSE | FALSE     | FALSE     |
NULL     | NULL          | Exam   | NULL        | FOUND       | ---   | ---       | ---       |
NULL     | NULL          | Exam   | NULL        | NOT FOUND   | TRUE  | TRUE      | TRUE      |
NULL     | NULL          | Exam   | PrairieTest | FOUND       | ---   | ---       | ---       |
NULL     | NULL          | Exam   | PrairieTest | NOT FOUND   | FALSE | TRUE      | FALSE     | Intended change for this PR
NULL     | NULL          | Exam   | Other       | FOUND       | ---   | ---       | ---       |
NULL     | NULL          | Exam   | Other       | NOT FOUND   | TRUE  | TRUE      | TRUE      |
NULL     | not-NULL      | Exam   | NULL        | FOUND       | FALSE | TRUE      | TRUE      | Do we want a change here?
NULL     | not-NULL      | Exam   | NULL        | NOT FOUND   | FALSE | FALSE     | TRUE      | Unintended change?
NULL     | not-NULL      | Exam   | PrairieTest | FOUND       | FALSE | TRUE      | TRUE      | Do we want a change here?
NULL     | not-NULL      | Exam   | PrairieTest | NOT FOUND   | FALSE | FALSE     | FALSE     |
NULL     | not-NULL      | Exam   | Other       | FOUND       | FALSE | TRUE      | TRUE      | Do we want a change here?
NULL     | not-NULL      | Exam   | Other       | NOT FOUND   | FALSE | FALSE     | TRUE      | Unintended change?
NULL     | NULL          | Public | NULL        | FOUND       | ---   | ---       | ---       |
NULL     | NULL          | Public | NULL        | NOT FOUND   | TRUE  | TRUE      | TRUE      |
NULL     | NULL          | Public | PrairieTest | FOUND       | ---   | ---       | ---       |
NULL     | NULL          | Public | PrairieTest | NOT FOUND   | ---   | ---       | ---       |
NULL     | NULL          | Public | Other       | FOUND       | ---   | ---       | ---       |
NULL     | NULL          | Public | Other       | NOT FOUND   | TRUE  | TRUE      | TRUE      |
NULL     | not-NULL      | Public | NULL        | FOUND       | FALSE | FALSE     | FALSE     |
NULL     | not-NULL      | Public | NULL        | NOT FOUND   | FALSE | FALSE     | FALSE     |
NULL     | not-NULL      | Public | PrairieTest | FOUND       | FALSE | FALSE     | FALSE     |
NULL     | not-NULL      | Public | PrairieTest | NOT FOUND   | FALSE | FALSE     | FALSE     |
NULL     | not-NULL      | Public | Other       | FOUND       | FALSE | FALSE     | FALSE     |
NULL     | not-NULL      | Public | Other       | NOT FOUND   | FALSE | FALSE     | FALSE     |
```